eventdisplay/tracks/FairEveTracks.h

```diff
@@ -0,0 +1,123 @@
+/********************************************************************************
+ *    Copyright (C) 2020 GSI Helmholtzzentrum fuer Schwerionenforschung GmbH    *
+ *                                                                              *
+ *              This software is distributed under the terms of the             *
+ *              GNU Lesser General Public Licence (LGPL) version 3,             *
+ *                  copied verbatim in the file "LICENSE"                       *
+ ********************************************************************************/
+/*
+ * FairEveTracks.h
+ *
+ *  Created on: 4 maj 2020
+ *      Author: Daniel Wielanek
+ *		E-mail: daniel.wielanek@gmail.com
+ *		Warsaw University of Technology, Faculty of Physics
+ */
+#ifndef FAIRTEVERACKS_H_
+#define FAIRTEVERACKS_H_
+#include "FairEventManager.h"
+#include "FairTask.h"
+
+#include <Rtypes.h>
+#include <RtypesCore.h>
+#include <TClonesArray.h>
+#include <TEveElement.h>
+#include <TEveTrack.h>
+#include <TEveTrackPropagator.h>
+#include <TGeoTrack.h>
+#include <TNamed.h>
+
+/**
+ * base evelement that represents tracks
+ */
+class FairEveTracks
+    : public TEveElement
+    , public TNamed
+{
+    FairEventManager *fEventManager;
+    TObjArray *fEveTrList;
+    Double_t fPt[2];
+    Double_t fEta[2];
+    Double_t fEnergy[2];
+    Bool_t fUsePt;
+    Bool_t fUseEta;
+    Bool_t fUseEnergy;
+
+  protected:
+    Double_t GetPtMin() const { return fPt[0]; };
+    Double_t GetPtMax() const { return fPt[1]; };
+    Double_t GetEtaMin() const { return fEta[0]; };
+    Double_t GetEtaMax() const { return fEta[1]; }
+    Double_t GetEnergyMin() const { return fEnergy[0]; };
+    Double_t GetEnergyMax() const { return fEnergy[1]; };
+    ;
+    Bool_t UsePtCut() const { return fUsePt; };
+    Bool_t UseEtaCut() const { return fUseEta; };
+    Bool_t UseEnergyCut() const { return fUseEnergy; };
+    TEveTrackPropagator *fTrPropagator;
+    TEveTrackList *fTrackGroup;
+    /**
+     * reset object group (track groups)
+     */
+    void ResetGroup();
+    FairEventManager *GetEventManager() const { return fEventManager; };
+    TObjArray *GetTracksList() const { return fEveTrList; };
+    /**
+     *
+     * @param tr
+     * @return track group based on parameter
+     */
+    virtual TEveTrackList *GetTrackGroup(void *tr);
```

Comment:
Now you added a base class implementation which does not depend on the `void*` argument at all? Again, what is the use case? Who is calling this function?

I was thinking more in a direction like this:
```diff
diff --git a/eventdisplay/tracks/FairEveGeoTracks.cxx b/eventdisplay/tracks/FairEveGeoTracks.cxx
index f32c89f57..2520afcd5 100644
--- a/eventdisplay/tracks/FairEveGeoTracks.cxx
+++ b/eventdisplay/tracks/FairEveGeoTracks.cxx
@@ -132,9 +132,8 @@ void FairEveGeoTracks::Repaint()
     gEve->Redraw3D(kFALSE);
 }

-TEveTrackList *FairEveGeoTracks::GetTrackGroup(void *tr)
+TEveTrackList *FairEveGeoTracks::GetTrackGroup(TGeoTrack *track)
 {
-    TGeoTrack *track = static_cast<TGeoTrack *>(tr);
     TParticle *P = static_cast<TParticle *>(track->GetParticle());
     fTrackGroup = nullptr;
     for (Int_t i = 0; i < GetTracksList()->GetEntriesFast(); i++) {
diff --git a/eventdisplay/tracks/FairEveGeoTracks.h b/eventdisplay/tracks/FairEveGeoTracks.h
index 9af902fd5..a127cf511 100644
--- a/eventdisplay/tracks/FairEveGeoTracks.h
+++ b/eventdisplay/tracks/FairEveGeoTracks.h
@@ -46,7 +46,7 @@ class FairEveGeoTracks : public FairEveTracks
     Bool_t CheckCuts(TGeoTrack *tr);
     void DrawTrack(Int_t id);
     void DrawAnimatedTrack(Int_t id);
-    TEveTrackList *GetTrackGroup(void *tr);
+    TEveTrackList *GetTrackGroup(TGeoTrack *tr);

   public:
     FairEveGeoTracks();
diff --git a/eventdisplay/tracks/FairEveTracks.cxx b/eventdisplay/tracks/FairEveTracks.cxx
index ecb5fd037..9f5762216 100644
--- a/eventdisplay/tracks/FairEveTracks.cxx
+++ b/eventdisplay/tracks/FairEveTracks.cxx
@@ -73,16 +73,3 @@ InitStatus FairEveTracks::Init()
 }

 FairEveTracks::~FairEveTracks() {}
-
-TEveTrackList *FairEveTracks::GetTrackGroup(void *tr)
-{
-    if (fTrackGroup == nullptr) {
-        fTrPropagator = new TEveTrackPropagator();
-        fTrackGroup = new TEveTrackList("tracks", fTrPropagator);
-        fTrackGroup->SetMainColor(kRed);
-        GetTracksList()->Add(fTrackGroup);
-        gEve->AddElement(fTrackGroup, this);
-        fTrackGroup->SetRnrLine(kTRUE);
-    }
-    return fTrackGroup;
-}
diff --git a/eventdisplay/tracks/FairEveTracks.h b/eventdisplay/tracks/FairEveTracks.h
index 60c53d294..167bacd06 100644
--- a/eventdisplay/tracks/FairEveTracks.h
+++ b/eventdisplay/tracks/FairEveTracks.h
@@ -62,12 +62,6 @@ class FairEveTracks
     void ResetGroup();
     FairEventManager *GetEventManager() const { return fEventManager; };
     TObjArray *GetTracksList() const { return fEveTrList; };
-    /**
-     *
-     * @param tr
-     * @return track group based on parameter
-     */
-    virtual TEveTrackList *GetTrackGroup(void *tr);

   public:
     FairEveTracks();
diff --git a/examples/advanced/propagator/src/FairEveRecoTracksExample.cxx b/examples/advanced/propagator/src/FairEveRecoTracksExample.cxx
index 642fe09d5..752835d36 100644
--- a/examples/advanced/propagator/src/FairEveRecoTracksExample.cxx
+++ b/examples/advanced/propagator/src/FairEveRecoTracksExample.cxx
@@ -160,9 +160,8 @@ void FairEveRecoTracksExample::DrawTrack(Int_t id)
     trList->AddElement(track);
 }

-TEveTrackList *FairEveRecoTracksExample::GetTrackGroup(void *tr)
+TEveTrackList *FairEveRecoTracksExample::GetTrackGroup(FairTutPropTrack *track)
 {
-    FairTutPropTrack *track = static_cast<FairTutPropTrack *>(tr);
     Color_t col = kYellow;
     TString pid = "neutral";
     Int_t q = track->GetParamFirst().GetQ();
diff --git a/examples/advanced/propagator/src/FairEveRecoTracksExample.h b/examples/advanced/propagator/src/FairEveRecoTracksExample.h
index 2d847e436..e031ce0a3 100644
--- a/examples/advanced/propagator/src/FairEveRecoTracksExample.h
+++ b/examples/advanced/propagator/src/FairEveRecoTracksExample.h
@@ -40,7 +40,7 @@ class FairEveRecoTracksExample : public FairEveTracks
   protected:
     Bool_t CheckCuts(FairTutPropTrack *tr);
     void DrawTrack(Int_t id);
-    TEveTrackList *GetTrackGroup(void *tr);
+    TEveTrackList *GetTrackGroup(FairTutPropTrack *tr);

   public:
     FairEveRecoTracksExample();
diff --git a/examples/common/eventdisplay/FairEveMCTracks.cxx b/examples/common/eventdisplay/FairEveMCTracks.cxx
index e2c67942e..a392a0c45 100644
--- a/examples/common/eventdisplay/FairEveMCTracks.cxx
+++ b/examples/common/eventdisplay/FairEveMCTracks.cxx
@@ -120,9 +120,8 @@ void FairEveMCTracks::DrawTrack(Int_t id)
     trList->AddElement(track);
 }

-TEveTrackList *FairEveMCTracks::GetTrackGroup(void *tr)
+TEveTrackList *FairEveMCTracks::GetTrackGroup(FairMCTrack *track)
 {
-    FairMCTrack *track = static_cast<FairMCTrack *>(tr);
     TString pid = Form("%i", track->GetPdgCode());
     fTrackGroup = nullptr;
     for (Int_t i = 0; i < GetTracksList()->GetEntriesFast(); i++) {
diff --git a/examples/common/eventdisplay/FairEveMCTracks.h b/examples/common/eventdisplay/FairEveMCTracks.h
index 34de1cd0d..439e5d1a6 100644
--- a/examples/common/eventdisplay/FairEveMCTracks.h
+++ b/examples/common/eventdisplay/FairEveMCTracks.h
@@ -35,7 +35,7 @@ class FairEveMCTracks : public FairEveTracks
   protected:
     Bool_t CheckCuts(FairMCTrack *tr);
     void DrawTrack(Int_t id);
-    TEveTrackList *GetTrackGroup(void *tr);
+    TEveTrackList *GetTrackGroup(FairMCTrack *tr);

   public:
     FairEveMCTracks();
diff --git a/examples/common/eventdisplay/FairMCTracksDraw.h b/examples/common/eventdisplay/FairMCTracksDraw.h
index 9b7db7e61..2b9666f88 100644
--- a/examples/common/eventdisplay/FairMCTracksDraw.h
+++ b/examples/common/eventdisplay/FairMCTracksDraw.h
@@ -28,7 +28,6 @@ class FairMCTracksDraw : public FairTask
   protected:
     Bool_t CheckCuts(FairMCTrack *tr);
     void DrawTrack(Int_t id);
-    TEveTrackList *GetTrackGroup(void *tr);

   public:
     FairMCTracksDraw();
```

If you cannot find a common type for the argument other than `void*`, why making it a virtual function? All your specializations static_cast it to a specific type unconditionally anyways - that means those are all just different specific function with nothing in common, right? What am I missing?

Reply:
Also refer to [I.4: Make interfaces precisely and strongly typed](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#i4-make-interfaces-precisely-and-strongly-typed)

Reply:
I make this only to mark that by default track list is cleaned up and user should check. That's why there is allocation of track group without type checking in FairEveTracks (that in previous version was pure virtual).
But if you think that this version is better - ok.

Reply:
Update: I think I found better solution - now this function is not virtual. User only in constructor of base class (FairEveTracks) define if he use TEveTrackList or FairEveTrackList (that accept any compound object).
Thanks to this he doesn't have to think about creation of track groups at all.

Reply:
> Update: I think I found better solution - now this function is not virtual. User only in constructor of base class (FairEveTracks) define if he use TEveTrackList or FairEveTrackList (that accept any compound object).
Thanks to this he doesn't have to think about creation of track groups at all.

I like it! Made some technical comments though.